### ISSUE TITLE
Validator manager contract utils

### ISSUE THREAD

**Issue by hwwhww** [ethereum/py-evm#197]
Implement some APIs for shard chain validators to call the `validator manager contract` on main chain.

* contract code: https://github.com/ethereum/sharding/blob/develop/sharding/contracts/validator_manager.v.py
* previous implementation (with pyethereum): https://github.com/ethereum/sharding/blob/develop/sharding/validator_manager_utils.py

Implementation should probably remain *low* level.  A simplistic `Contract` class with the following API should suffice.

* Base class with `factory` class method which accepts `abi` and returns a `ContractClass`
* `ContractClass` as returned by `BaseContract.factory` can be instantiated with a single `address`.
* A method on the `ContractClass` for constructing a transaction (TODO: should this be a dictionary, a `Message` class, or something else?)

There are likely some other convenience APIs that should be added around this functionality to make common tasks easier as we find what those are.

I think this code should live in something like `evm.utils.contracts`.

**Comment by pipermerriam** [comment]
Made some updates to try and add some clarity to how this might be implemented.

**Comment by pipermerriam** [comment]
@carver looks like @mhchia may be working on this but you may be well suited to assist with implementation or just sussing out the API.

**Comment by carver** [comment]
Interesting, the contract API starts to feel a lot like web3.py: http://web3py.readthedocs.io/en/latest/web3.eth.html#web3.eth.Eth.contract 

Maybe we could rip out some part of web3 into a `eth-contracts` library, for code reuse. Either way, I'm happy to chat about API. As you can see in the docs, web3.py uses a `dict` when constructing a transaction.

**Comment by pipermerriam** [comment]
I'm going to backpedal hard on my original suggestion since I think I may have missed something crutial.

@mhchia does this interaction happen over the JSON-RPC interface?

**Comment by mhchia** [comment]
@pipermerriam 
Yes it is. Previously I used the `web3.py` contract object.
But after, I changed to generate the tx data of the contract call transaction on our own, using `eth_abi`. (referenced from [here](https://github.com/ethereum/eth-tester/blob/99dfc5a8f39988316f6bd6ede3c6942d8a2bb389/eth_tester/utils/math_contract.py))

Now performing a contract call looks like [this](https://github.com/mhchia/sharding/blob/rpc/sharding/rpc/vmc_handler.py#L101-L108), specifying the `function_name` and `args`.
The implementation underneath: [`VMCHandler.send_vmc_tx`](https://github.com/mhchia/sharding/blob/rpc/sharding/rpc/vmc_handler.py#L62-L73).
I did this is because I used `web3.py` and `eth_tester` as the backends of `vmc_handler`. The parameters taken for the `sendTransaction` of both packages are the same.
So I can just generate one parameter for both backends using the same function [`mk_contract_tx_obj`](https://github.com/mhchia/sharding/blob/rpc/sharding/rpc/vmc_utils.py#L152-L174).

If it is better to use web3.py contract object, I can change to that:)
But this way I will need to figure out a prettier solution.

**Comment by pipermerriam** [comment]
@mhchia 

Just in case you are not aware.  You may be able to slim down to a single handler that *only* uses web3.py and still leverage `eth-tester`.  Web3.py ships with a special provider `web3.providers.eth_tester.EthereumTesterProvider` which uses `eth-tester` under the hood.  That would allow a single handler implementation that just delegates to `web3.py` or potentially for you to remove the chain_handler all-together.

**Comment by mhchia** [comment]
@pipermerriam Thank you. It's so great! I will modify them to use `web3.py` altogether.

**Comment by mhchia** [comment]
I tried `EthereumTesterProvider` with `EthereumTester(PyEVMBackend())`, however it seems there are still some of the RPCs not implemented. May I change to use that later when  `EthereumTesterProvider` is more complete? I think I can help with that after the https://github.com/ethereum/py-evm/pull/206 . Or it is also fine that we can do it right away and continue to https://github.com/ethereum/py-evm/pull/206 ?
Ah I'm in a hurry for this part(Since `TesterChainHandler` is currently enough for `vmc_handler` to work. Please reply only when you have spare time. I really appreciate your help! :D

**Comment by pipermerriam** [comment]
@mhchia which endpoints?  Depending on what they are it should be easy to implement them.

**Comment by mhchia** [comment]
I'm really sorry that I just found I missed a word in the comment above(*I'm not in a hurry for this part...). I apologize if I made you misunderstand.
About the endpoints, it seems like `miner.start()` is not implemented yet in the [eth_tester provider](https://github.com/ethereum/web3.py/blob/master/web3/providers/eth_tester/main.py).
I will fix the endpoints that is needed and use eth-tester provider and remove the `TesterChainHandler`. Thank you!
